packages/flutter/lib/src/widgets/change_notifier_builder.dart

```diff
@@ -0,0 +1,151 @@
+// found in the LICENSE file.
+
+import 'basic.dart';
+import 'framework.dart';
+
+/// Builds a [Widget] when given a concrete value of a [ChangeNotifier].
+///
+/// If the `child` parameter provided to the [ChangeNotifierBuilder] is not
+/// null, the same `child` widget is passed back to this
+/// [ChangeNotifierWidgetBuilder] and should typically be incorporated in the
+/// returned widget tree.
+///
+/// See also:
+///
+///  * [ChangeNotifierBuilder], a widget which invokes this builder each time
+///    a [ChangeNotifier] changes value.
+typedef ChangeNotifierWidgetBuilder<T> = Widget Function(BuildContext context, Widget? child);
+
+/// A widget whose content stays synced with a [ChangeNotifier].
+///
+/// Given a [ChangeNotifier] and a [builder] which builds widgets from
+/// subtypes of `ChangeNotifier`, this class will automatically register itself
+/// as a listener of the [ChangeNotifier] and call the [builder] whenever
+/// listeners are notified.
+///
+/// An [AnimatedBuilder] could be used instead of this widget but it doesn't
+/// have a [listenCondition] callback and it's not specialized to only listen to
+/// [ChangeNotifier]s.
+///
+/// ## Performance optimizations
+///
+/// If your [builder] function contains a subtree that does not depend on the
+/// value of the [ChangeNotifier], it's more efficient to build that subtree
+/// once instead of rebuilding it on every listener notification.
+///
+/// If you pass the pre-built subtree as the [child] parameter, the
+/// [ChangeNotifierBuilder] will pass it back to your [builder] function so
+/// that you can incorporate it into your build.
+///
+/// Using this pre-built child is entirely optional, but can improve
+/// performance significantly in some cases and is therefore a good practice.
+///
+/// See also:
+///
+///  * [AnimatedBuilder], which can trigger rebuilds from a [Listenable].
+///  * [NotificationListener], which lets you rebuild based on [Notification]
+///    coming from its descendant widgets rather than a [ValueListenable] that
+///    you have a direct reference to.
+///  * [ValueListenableBuilder], which can trigger rebuilds from a
+///  [ValueNotifier].
+class ChangeNotifierBuilder<T extends ChangeNotifier> extends StatefulWidget {
+  /// Creates a widget that rebuilds whenever the given change notifier notifies
+  /// its listeners.
+  ///
+  /// The [changeNotifier] and [builder] arguments are required.
+  ///
+  /// The [child] is optional but is good practice to use if part of the widget
+  /// subtree does not depend on the value of the [changeNotifier].
+  ///
+  /// The [listenCondition] is also optional and it can be used to control when
+  /// the [builder] should be called.
+  const ChangeNotifierBuilder({
+    Key? key,
+    required this.changeNotifier,
+    required this.builder,
+    this.listenCondition,
+    this.child,
+  }) : super(key: key);
+
+  /// The [ChangeNotifier] whose value you depend on in order to build.
+  final T changeNotifier;
```

Comment:
Why is this typed as T instead of just "ChangeNotifier" (or really, "Listenable")?